File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/MultipartPathOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.MultipartPathOverlay
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler.Companion.getterName
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NMultipartPathOverlay
import io.flutter.plugin.common.MethodChannel

internal interface MultipartPathOverlayHandler : OverlayHandler {
    fun handleMultipartPathOverlay(
        multipartPathOverlay: Overlay,
        method: String,
        arg: Any?,
        result: MethodChannel.Result,
    ) = (multipartPathOverlay as MultipartPathOverlay).let { m ->
        when (method) {
            NMultipartPathOverlay.pathsName -> setPaths(m, arg!!)
            NMultipartPathOverlay.widthName -> setWidth(m, arg!!)
            NMultipartPathOverlay.outlineWidthName -> setOutlineWidth(m, arg!!)
            NMultipartPathOverlay.patternImageName -> setPatternImage(m, arg!!)
            NMultipartPathOverlay.patternIntervalName -> setPatternInterval(m, arg!!)
            NMultipartPathOverlay.progressName -> setProgress(m, arg!!)
            NMultipartPathOverlay.isHideCollidedCaptionsName -> setIsHideCollidedCaptions(m, arg!!)
            NMultipartPathOverlay.isHideCollidedMarkersName -> setIsHideCollidedMarkers(m, arg!!)
            NMultipartPathOverlay.isHideCollidedSymbolsName -> setIsHideCollidedSymbols(m, arg!!)
            getterName(NMultipartPathOverlay.boundsName) -> getBounds(m, result::success)
            else -> result.notImplemented()
        }
    }

    fun setPaths(multipartPathOverlay: MultipartPathOverlay, rawPaths: Any)

    fun setWidth(multipartPathOverlay: MultipartPathOverlay, rawWidthDp: Any)

    fun setOutlineWidth(multipartPathOverlay: MultipartPathOverlay, rawWidthDp: Any)

    fun setPatternImage(multipartPathOverlay: MultipartPathOverlay, rawNOverlayImage: Any)

    fun setPatternInterval(multipartPathOverlay: MultipartPathOverlay, rawInterval: Any)

    fun setProgress(multipartPathOverlay: MultipartPathOverlay, rawProgress: Any)

    fun setIsHideCollidedCaptions(multipartPathOverlay: MultipartPathOverlay, rawFlag: Any)

    fun setIsHideCollidedMarkers(multipartPathOverlay: MultipartPathOverlay, rawFlag: Any)

    fun setIsHideCollidedSymbols(multipartPathOverlay: MultipartPathOverlay, rawFlag: Any)

    fun getBounds(
        multipartPathOverlay: MultipartPathOverlay,
        success: (bounds: Map<String, Any>) -> Unit,
    )
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/MultipartPathOverlayHandler.swift
import NMapsMap
import Flutter

internal protocol MultipartPathOverlayHandler: OverlayHandler {
    func setPaths(_ multipartPathOverlay: NMFMultipartPath, rawPaths: Any)

    func setWidth(_ multipartPathOverlay: NMFMultipartPath, rawWidthDp: Any)

    func setOutlineWidth(_ multipartPathOverlay: NMFMultipartPath, rawWidth: Any)

    func setPatternImage(_ multipartPathOverlay: NMFMultipartPath, rawNOverlayImage: Any)

    func setPatternInterval(_ multipartPathOverlay: NMFMultipartPath, rawInterval: Any)

    func setProgress(_ multipartPathOverlay: NMFMultipartPath, rawProgress: Any)

    func setIsHideCollidedCaptions(_ multipartPathOverlay: NMFMultipartPath, rawFlag: Any)

    func setIsHideCollidedMarkers(_ multipartPathOverlay: NMFMultipartPath, rawFlag: Any)

    func setIsHideCollidedSymbols(_ multipartPathOverlay: NMFMultipartPath, rawFlag: Any)

    func getBounds(_ multipartPathOverlay: NMFMultipartPath, success: (_ bounds: Dictionary<String, Any>) -> Void)
}

internal extension  MultipartPathOverlayHandler {
    func handleMultipartPathOverlay(multipartPathOverlay: NMFOverlay,
                                    method: String,
                                    args: Any?,
                                    result: @escaping FlutterResult) {
        let multipartPathOverlay = multipartPathOverlay as! NMFMultipartPath
        switch method {
        case NMultipartPathOverlay.pathsName: setPaths(multipartPathOverlay, rawPaths: args!)
        case NMultipartPathOverlay.widthName: setWidth(multipartPathOverlay, rawWidthDp: args!)
        case NMultipartPathOverlay.outlineWidthName: setOutlineWidth(multipartPathOverlay, rawWidth: args!)
        case NMultipartPathOverlay.patternImageName: setPatternImage(multipartPathOverlay, rawNOverlayImage: args!)
        case NMultipartPathOverlay.patternIntervalName: setPatternInterval(multipartPathOverlay, rawInterval: args!)
        case NMultipartPathOverlay.progressName: setProgress(multipartPathOverlay, rawProgress: args!)
        case NMultipartPathOverlay.isHideCollidedCaptionsName: setIsHideCollidedCaptions(multipartPathOverlay, rawFlag: args!)
        case NMultipartPathOverlay.isHideCollidedMarkersName: setIsHideCollidedMarkers(multipartPathOverlay, rawFlag: args!)
        case NMultipartPathOverlay.isHideCollidedSymbolsName: setIsHideCollidedSymbols(multipartPathOverlay, rawFlag: args!)
        case getterName(NMultipartPathOverlay.boundsName): getBounds(multipartPathOverlay, success: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
